### PR TITLE
Feature/1838 cache post requests

### PR DESCRIPTION
**Pull request by komalsukhani**
Fix issue #1838 

A new field `advance_cache_config` is added in API definition. It has following format:
```json
"advance_cache_config": [
				{
					"method":"POST",
					"path":"addBooks",
					"cache_key_regex": "pattern"
				}
			]
```
 If user wants to enable caching of POST for specific path, an entry should be added in `advance_cache_config`. 

`cache_key_regex` enables user to specify regexp to extract cache key from the body. If it set to "", entire post body will be used to generate cache key. `method` field enables caching of request for specific HTTP method. `cache_key_regex` will not be used for safe methods(GET/HEADERS/OPTIONS). It's value is ignored.

### CODE REVIEW

**Comment by buger** [comment]
Thank you! Going to review soon.

File: mw_redis_cache.go [hunk 1]
```diff
@@ -41,13 +44,42 @@ func (m *RedisCacheMiddleware) EnabledForSpec() bool {
 	return m.Spec.CacheOptions.EnableCache
 }
 
-func (m *RedisCacheMiddleware) CreateCheckSum(req *http.Request, keyName string) string {
+func (m *RedisCacheMiddleware) CreateCheckSum(req *http.Request, keyName string, regex string) (string, error) {
 	h := md5.New()
 	io.WriteString(h, req.Method)
 	io.WriteString(h, "-")
 	io.WriteString(h, req.URL.String())
+	if req.Method == "POST" {
```
Comment (asoorm) [inline mw_redis_cache.go]: I would prefer to use package contstant `http.MethodPost`

File: mw_redis_cache.go [hunk 2]
```diff
@@ -41,13 +44,42 @@ func (m *RedisCacheMiddleware) EnabledForSpec() bool {
...
+		if req.Body != nil {
+			bodyBytes, err := ioutil.ReadAll(req.Body)
+			defer req.Body.Close()
```
Comment (asoorm) [inline mw_redis_cache.go]: `defer req.Body.Close()` should be called after `if err != nil {...}`

Comment (komalsukhani) [reply mw_redis_cache.go]: Oh yeah. I will fix it.

File: mw_redis_cache.go (hunk 1 above)
Comment (komalsukhani) [reply mw_redis_cache.go]: ok

File: gateway_test.go [hunk 3]
```diff
@@ -1032,6 +1032,39 @@ func TestCacheAllSafeRequests(t *testing.T) {
...
+                                ]`), &v.ExtendedPaths.AdvanceCacheConfig)
+		})
+
+		spec.Proxy.ListenPath = "/"
+	})
+
+	headerCache := map[string]string{"x-tyk-cached-response": "1"}
+
+	ts.Run(t, []test.TestCase{
+		{Method: "POST", Path: "/", HeadersNotMatch: headerCache, Delay: 10 * time.Millisecond},
```
Comment (joshblakeley) [inline gateway_test.go]: In this unit test are you actually sending a body with the post request?

Comment (joshblakeley) [reply gateway_test.go]: We should probably have a case testing the regex key against specific body.

Comment (komalsukhani) [reply gateway_test.go]: Fixed test cases.

**Pull request merged**